neres/newrelic.py

```diff
@@ -0,0 +1,241 @@
+    response.raise_for_status()
+    data = response.json()
+
+    # sort data by name
+    data = sorted(data, key=lambda x: x['name'])
+
+    return data
+
+
+@requires_login
+def delete_monitor(monitor):
+    response = session.get(urls.MONITOR_JSON.format(monitor))
+    response.raise_for_status()
+    headers = {
+        'Referer': response.url,
+    }
+
+    response = session.delete(urls.MONITOR_JSON.format(monitor),
+                              headers=headers)
+    response.raise_for_status()
+
+
+@requires_login
+def get_locations():
+    response = session.get(urls.MONITOR_LOCATIONS)
+    response.raise_for_status()
+
+    return response.json()
+
+
+@requires_login
+def get_monitor(monitor):
+    response = session.get(urls.MONITOR_JSON.format(monitor))
+    response.raise_for_status()
+
+    return response.json()
+
+
+@requires_login
+def update_monitor(monitor, *args, **kwargs):
+    data = get_monitor(monitor)
+    for prop in ['name', 'uri', 'frequency', 'locations', 'slaThreshold']:
+        value = kwargs.get(prop, None)
+        if value:
+            data[prop] = value
+
+    for prop in ['emails', 'locations']:
+        if kwargs.get('clear_{}'.format(prop)):
+            data[prop] = []
+
+        values = kwargs.get('add_{}'.format(prop))
+        if values:
+            if isinstance(values, six.string_types):
+                values = [values]
+            data[prop] = list(set(data[prop] + list(values)))
+
+        values = kwargs.get('remove_{}'.format(prop))
+        if values:
+            if isinstance(values, six.string_types):
+                prop = [values]
+            data[prop] = list(set(data[prop]) - set(values))
+
+    _construct_metadata(
+        validation_string=kwargs.get('validation_string'),
+        bypass_head_request=kwargs.get('bypass_head_request'),
+        verify_ssl=kwargs.get('verify_ssl'),
+        redirect_is_failure=kwargs.get('redirect_is_failure'),
+        metadata=data['metadata'])
+
+    headers = {
+        'Referer': urls.MONITOR.format(monitor),
+        'Content-Type': 'application/json;charset=utf-8',
+    }
+
+    url = urls.MONITOR_JSON.format(monitor)
+    response = session.put(url, data=json.dumps(data), headers=headers)
+
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        try:
+            error = response.json()['error']
+        except (ValueError, KeyError):
+            error = 'Bad request'
+        return (1, error, {})
+
+    return (0, urls.MONITOR.format(response.json()['id']), response.json())
+
+
+def _construct_metadata(validation_string=None, bypass_head_request=None,
+                        verify_ssl=None, redirect_is_failure=None,
+                        metadata=None):
+    if metadata is None:
+        metadata = {}
+
+    if bypass_head_request is True:
+        metadata['nr.synthetics.metadata.job.options.simple.bypass.head'] = True
+    elif bypass_head_request is False:
```

Comment:
bypass_head_request can be `None`, in the case `update` where I don't want to touch it